File: python/minesweeper.py
def calc_mines(minefield, i, j):
    mines = 0;
    rows = len(minefield)
    cols = len(minefield[0])
    for di in [-1, 0, 1]:
        for dj in [-1, 0, 1]:
            if i+di < 0 or i+di >= rows or j+dj < 0 or j+dj >= cols:
                continue
            if minefield[i+di][j+dj] == '*':
                mines += 1
    return mines


def annotate(minefield):
    if minefield == []:
        return []
        
    rows = len(minefield)
    cols = len(minefield[0])

    res = [[None] * cols for _ in range(rows)]
    
    for i in range(rows):
        if len(minefield[i]) != cols:
                raise ValueError('The board is invalid with current input.')
        for j in range(cols):
            if minefield[i][j] == '*':
                res[i][j] = '*'
            elif minefield[i][j] == ' ':
                mines = calc_mines(minefield, i, j)
                res[i][j] = mines == 0 and ' ' or str(mines)
            else:
                raise ValueError('The board is invalid with current input.')

    return [''.join(row) for row in res]

File: elixir/minesweeper.ex
defmodule Minesweeper do
  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """
  @spec annotate([String.t()]) :: [String.t()]
  def annotate([]), do: []

  def annotate(board) do
    row = length(board)
    col = String.length(hd(board))

    mp =
      board
      |> Enum.with_index(1)
      |> Enum.reduce(%{}, fn {r, i}, acc ->
        r
        |> String.graphemes()
        |> Enum.with_index(1)
        |> Enum.reduce(acc, fn {c, j}, acc ->
          Map.put(acc, {i, j}, c)
        end)
      end)

    mp
    |> Map.new(fn
      {k, "*"} ->
        {k, "*"}

      {k = {x, y}, _} ->
        [{-1, -1}, {-1, 0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}]
        |> Enum.count(fn {dx, dy} ->
          mp[{x + dx, y + dy}] == "*"
        end)
        |> case do
          0 -> " "
          n -> to_string(n)
        end
        |> then(&{k, &1})
    end)
    |> then(fn mp ->
      for i <- 1..row do
        for j <- 1..col do
          mp[{i, j}]
        end
        |> Enum.join("")
      end
    end)
  end
end
